strawberryfields/program.py

```diff
@@ -543,9 +591,15 @@ def _get_compiler(compiler_or_name):
             else:
                 compiler = _get_compiler(compiler)
 
-            # TODO: add validation for device specs that provide a dictionary for `device.modes`.
-            if device.modes is not None and isinstance(device.modes, int):
-                self.assert_number_of_modes(device)
+            if device.modes is not None:
+                if isinstance(device.modes, int):
+                    # Check that the number of modes is correct, if device.modes
+                    # is provided as an integer
+                    self.assert_number_of_modes(device)
+                elif device.target == "simulon":
```

Comment:
It's as if we'd prefer not to use explicit target names here, but cannot quite remember why (perhaps was just due to the fact that it could break easily) 🤔 @josh146? Perhaps it's also less relevant now that we'll have API versioning.

Reply:
I think it's because names may change. 

In this case, it's not that crucial though because of, as you mentioned, the API versioning. This part would need to be updated anyway as soon as we decide to use API v0.3.0, so for now I think it's fine to make sure that only Simulon calls `assert_num_of_measurments`, since it's the only target that actually has the correct type of modes entry in API v.0.2.0.

Reply:
For now, you could just have an `else:` here, and use it to just catch _everything?_ assuming TDM compilation does not go via this pipeline.

Reply:
Yes, that'd work as well. @jswinarton mentioned that the name will likely change in the future, so it's probably best to remove the hard-coded "simulon" from here.